app/services/distance_calculator.rb

```diff
@@ -0,0 +1,30 @@
+class DistanceCalculator
+  def initialize(geocoder, start_address, destination_address)
+    @geocoder = geocoder
+    @start_address = start_address
+    @destination_address = destination_address
+  end
+
+  def call
+    start_point = geocoder.geocode(start_address)
+    end_point = geocoder.geocode(destination_address)
+
+    if start_point.success && end_point.success
+      start_point.distance_from(end_point, units: :meters)
+    else
+      raise AddressNotFound, generate_error_messages(start_point, end_point)
+    end
+  end
+
+  private
+
+  def generate_error_messages(start_point, end_point)
+    messages = {}
+    messages.merge!(start_address: ['Could not find address']) unless start_point.success
+    messages.merge!(destination_address: ['Could not find address']) unless end_point.success
+
+    [messages]
```

Comment:
You could initialize empty array here and think about adding errors to it instead of using empty hash, merging sth to it and then wrapping it into the array. It looks kinda weird now.

Reply:
changed to array of hashes